src/Starlight/Physics.hs

```diff
@@ -40,17 +40,19 @@ import Unit.Length
 import Unit.Mass
 import Unit.Time
 
-inertia :: Has (Reader (Seconds Float)) sig m => Actor -> m Actor
-inertia a@Actor{ position, velocity } = do
+inertia :: (Has (Reader (Seconds Float)) sig m, HasCallStack) => Actor -> m Actor
+inertia a@Actor{ velocity } = do
   dt <- ask @(Seconds _)
-  pure a { Actor.position = position + ((.*. dt) <$> velocity) }
+  pure $! a & position_ +~ ((.*. dt) <$> velocity)
 
-gravity :: (Has (Reader (Seconds Float)) sig m, Has (Reader (System StateVectors)) sig m) => Actor -> m Actor
+gravity :: (Has (Reader (Seconds Float)) sig m, Has (Reader (System StateVectors)) sig m, HasCallStack) => Actor -> m Actor
 gravity a = do
   dt <- ask @(Seconds Float)
   System{ bodies } <- ask
   pure $! foldl' (go dt) a bodies where
-  go dt a StateVectors{ actor = b } = applyForce (force *^ coerce ((b^.position_) `direction` (a^.position_))) dt a where
+  go dt a StateVectors{ actor = b }
+    | nearZero r = a
```

Comment:
We check that the actor & body aren’t colocated so that we don’t end up with a 0 denominator.